File: python/ray/_private/runtime_env_prototype/sdk/java/src/test/java/io/ray/RuntimeEnvTest.java
package io.ray;

import java.util.Arrays;
import static org.junit.Assert.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.github.fge.jackson.JsonLoader;
import org.junit.Test;

public class RuntimeEnvTest
{
    @Test
    public void testRuntimeEnv() throws Exception {
      // Load plugin schemas
      final String currentDir = System.getProperty("user.dir");
      String[] schema_paths={currentDir + "/../../pip/pip_schema.json",
        currentDir + "/../../working_dir/working_dir_schema.json"};
      PluginSchemaManager.getInstance().loadSchemas(schema_paths);
      RuntimeEnv runtimeEnv = new RuntimeEnv();
      // Add pip runtime env
      Pip pip = new Pip();
      String[] packages = new String[] {"requests"};
      pip.packages = packages;
      pip.pip_check = true;
      runtimeEnv.set("pip", pip);
      // Add working dir runtime env
      String workingDir = "https://path/to/working_dir.zip";
      runtimeEnv.set("working_dir", workingDir);
      // Serialize
      String serializedRuntimeEnv = runtimeEnv.serialize();
      System.out.println("serializedRuntimeEnv " + serializedRuntimeEnv);

      // Deserialize
      RuntimeEnv runtimeEnv2 = RuntimeEnv.deserialize(serializedRuntimeEnv);
      Pip pip2 = runtimeEnv2.get("pip", Pip.class);
      assertTrue(Arrays.equals(pip2.packages, packages));
      String workingDir2 = runtimeEnv2.get("working_dir", String.class);
      assertTrue(workingDir2.equals(workingDir));

      // Construct runtime env with raw json
      RuntimeEnv runtimeEnv3 = new RuntimeEnv();
      String pipString = "    "
        + "{\n"
        + "      \"packages\": [\"requests\", \"tensorflow\"],\n"
        + "      \"pip_check\": false\n"
        + "}";
      JsonNode pipRawJson = JsonLoader.fromString(pipString);
      runtimeEnv3.set("pip", pipRawJson);
      System.out.println("serializedRuntimeEnv 3: " + runtimeEnv3.serialize());

      // Construct runtime env with raw json string
      RuntimeEnv runtimeEnv4 = new RuntimeEnv();
      runtimeEnv4.set("pip", pipString);
      System.out.println("serializedRuntimeEnv 4: " + runtimeEnv4.serialize());
    }
}

File: python/ray/_private/runtime_env_prototype/sdk/cpp/runtime_env_test.cc
#include <iostream>
#include "pip.h"
#include "runtime_env.h"
#include "plugin_schema_manager.h"
#include "nlohmann/json.hpp"
using json = nlohmann::json;

int main(int argc, char **argv) {
  // Load plugin schemas
  std::vector<std::string> schema_paths = {
    "/home/admin/ray/python/ray/_private/runtime_env_prototype/pip/pip_schema.json",
    "/home/admin/ray/python/ray/_private/runtime_env_prototype/working_dir/working_dir_schema.json"
  };
  RuntimeEnvPluginSchemaManager::GetInstance().LoadSchemas(schema_paths);

  RuntimeEnv runtime_env;
  // Set pip
  std::vector<std::string> packages = {"requests"};
  Pip pip(packages, true);
  runtime_env.Set("pip", pip);
  // Set working_dir
  std::string working_dir = "https://path/to/working_dir.zip";
  runtime_env.Set("working_dir", working_dir);

  // Serialize
  auto serialized_runtime_env = runtime_env.Serialize();
  std::cout << "serialized_runtime_env " << serialized_runtime_env << std::endl;

  // Deserialize
  auto runtime_env_2 = RuntimeEnv::Deserialize(serialized_runtime_env);

  auto pip2 = runtime_env_2.Get<Pip>("pip");
  assert(pip2.packages == pip.packages);
  assert(pip2.pip_check == pip.pip_check);

  auto working_dir2 = runtime_env_2.Get<std::string>("working_dir");
  assert(working_dir2 == working_dir);

  // Construct runtime env with raw json string
  RuntimeEnv runtime_env_3;
  std::string pip_raw_json_string = R"({"packages":["requests","tensorflow"],"pip_check":false})";
  runtime_env_3.SetJsonStr("pip", pip_raw_json_string);
  std::cout << "serialized_runtime_env 3: " << runtime_env_3.Serialize() << std::endl;
  auto get_json_result = runtime_env_3.GetJsonStr("pip");
  assert(get_json_result == pip_raw_json_string);

  std::cout << "Finished!" << std::endl;;

  return 0;
}
